core/core-api/src/main/java/io/fabric8/launcher/core/spi/IdentityProviders.java

```diff
@@ -0,0 +1,57 @@
+package io.fabric8.launcher.core.spi;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import io.fabric8.launcher.base.identity.Identity;
+import io.fabric8.launcher.base.identity.ImmutableTokenIdentity;
+import io.fabric8.launcher.base.identity.TokenIdentity;
+
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * An utility class for {@link IdentityProviders}.
+ */
+public final class IdentityProviders {
+
+    private IdentityProviders() {
+        throw new IllegalAccessError("utility class");
+    }
+
+    public static Optional<Identity> useDefaultOrElseProvide(final Optional<Identity> defaultIdentity, final IdentityProvider identityProvider, final TokenIdentity authorization, final String provider) {
+        if (defaultIdentity.isPresent()) {
+            return defaultIdentity;
+        } else {
+            return identityProvider.getIdentity(authorization, provider);
+        }
+    }
+
+    public static Map<String, Optional<Identity>> getIdentities(final IdentityProvider identityProvider, final TokenIdentity authorization, final Set<String> services) {
```

Comment:
This code looks really complicated. Are you sure we need this?

Reply:
Do you see an alternative? or a better way? (look how it used)

Reply:
The issue is that you don't want to loose providerName for the identities